### PR TITLE
prepareSearchQuery callable function

### PR DESCRIPTION
**Pull request by programmis**
* prepareSearchQuery callable function in IndexAction

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

### CODE REVIEW

**Comment by bizley** [comment]
There is already `prepareDataProvider` with which you can modify search query. Why do you think additional property is needed here?

**Comment by programmis** [comment]
> There is already `prepareDataProvider` with which you can modify search query. Why do you think additional property is needed here?

to change exactly the query. For example, if you need to work with linked tables

**Comment by bizley** [comment]
@samdark what do you think about this?

**Comment by samdark** [comment]
> For example, if you need to work with linked tables

@programmis would you please elaborate about the use case?

**Comment by programmis** [comment]
> > For example, if you need to work with linked tables
> 
> @programmis would you please elaborate about the use case?

```
class MyController extends \yii\rest\ActiveController {
    /** @inheritdoc */
    public function actions()
    {
        return ArrayHelper::merge(parent::actions(), [
            'index' => [
                'dataFilter' => [
                    'class' => ActiveDataFilter::class,
                    //I would search by field in join table
                    'attributeMap' => ['positionIds' => LinkConditionToPosition::tableName() . '.[[positionId]]'],
                ],
                'prepareSearchQuery' => function ($query, $requestParams) {
                    if (ArrayHelper::getValue($requestParams, 'filter.positionIds')) {
                        //And now it is easy to do fast. Just do like this with prepared query
                        $query->leftJoin(
                            LinkConditionToPosition::tableName(),
                            LinkConditionToPosition::tableName() . '.[[conditionId]] = ' . Condition::tableName() . '.[[id]]'
                        );
                    }

                    return $query;
                },
            ],
        ]);
    }
}
```

File: framework/rest/IndexAction.php [hunk 1]
```diff
@@ -45,6 +45,16 @@ class IndexAction extends Action
      * ```
      */
     public $prepareDataProvider;
+    /**
+     * ```php
```
Comment (samdark) [inline framework/rest/IndexAction.php:49]: Need `@since 2.0.41`.

**Review by samdark — APPROVED** [review]
Would you please add a line for CHANGELOG.md and some unit tests? Thanks.

File: framework/rest/IndexAction.php [hunk 2]
```diff
@@ -115,6 +131,9 @@ protected function prepareDataProvider()
         if (!empty($filter)) {
             $query->andWhere($filter);
         }
+        if ($this->prepareSearchQuery && is_callable($this->prepareSearchQuery)) {
```
Comment (bizley) [inline framework/rest/IndexAction.php:134]:
```suggestion
        if (is_callable($this->prepareSearchQuery)) {
```

File: tests/framework/rest/IndexActionTest.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
+<?php
+
+
```
Comment (bizley) [inline tests/framework/rest/IndexActionTest.php:3]: Remove empty line

File: tests/framework/rest/IndexActionTest.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+namespace yiiunit\framework\rest;
+
```
Comment (bizley) [inline tests/framework/rest/IndexActionTest.php:5]: Remove empty line

**Review by bizley — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Pull request merged**